src/main/java/com/mongodb/jdbc/oidc/OidcAuthFlow.java

```diff
@@ -54,13 +56,58 @@ public class OidcAuthFlow {
     private static final Logger logger = Logger.getLogger(OidcAuthFlow.class.getName());
     private MongoLogger mongoLogger;
     private static final String OFFLINE_ACCESS = "offline_access";
+    private static final String OPENID = "openid";
 
     public OidcAuthFlow() {}
 
     public OidcAuthFlow(MongoLogger parentLogger) {
         this.mongoLogger = new MongoLogger(OidcAuthFlow.class.getName(), parentLogger);
     }
 
+    /**
+     * Builds the OIDC scopes for the authorization request by combining default scopes with
+     * client-requested scopes that are supported by the Identity Provider. Adds openid and
+     * offline_access scopes by default, and includes the clientID/.default if in the requested
+     * scopes.
+     *
+     * @param clientID the OAuth2 client identifier
+     * @param idpServerInfo server information containing requested scopes and other IdP details
+     * @param providerMetadata OIDC provider metadata containing supported scopes
+     * @return a Scope object containing all valid scopes to be requested
+     */
+    public Scope buildScopes(
+            String clientID, IdpInfo idpServerInfo, OIDCProviderMetadata providerMetadata) {
+        Set<String> scopes = new HashSet<>();
+        Scope supportedScopes = providerMetadata.getScopes();
+
+        // Add openid and offline_access scopes by default
+        scopes.add(OPENID);
+        scopes.add(OFFLINE_ACCESS);
+
+        // Add custom scopes from request that are supported by the IdP
+        List<String> requestedScopes = idpServerInfo.getRequestScopes();
+        if (requestedScopes != null) {
+            // Always add clientID/.default if it's in requestedScopes, this was needed for Azure OIDC.
+            String clientIDDefault = clientID + "/.default";
+            if (requestedScopes.contains(clientIDDefault)) {
+                scopes.add(clientIDDefault);
+            }
+            for (String scope : requestedScopes) {
+                if (supportedScopes != null) {
+                    if (supportedScopes.contains(scope)) {
+                        scopes.add(scope);
+                    }
+                }
```

Comment:
Let's keep the previous logic with the warning, since we can potentially "ditch" scopes.
```suggestion
if (supportedScopes != null && supportedScopes.contains(scope)) {
                        scopes.add(new Scope.Value(scope));
                    } else {
                        logger.warning(
                                "Requested scope '" + scope + "' is not supported by the IdP");
                    }
```

Reply:
feels a bit odd to check null every time, but as you mentioned this won't be a long loop!

Reply:
Oh, good catch on the warning I re-added it.  
I also moved the null check to outside of the loop.